src/jit/emitxarch.cpp

```diff
@@ -105,9 +183,22 @@ bool emitter::IsDstDstSrcAVXInstruction(instruction ins)
 // to indicate whether a 3-operand instruction.
 bool emitter::IsDstSrcSrcAVXInstruction(instruction ins)
 {
-    return IsAVXInstruction(ins) &&
-           (ins == INS_movlpd || ins == INS_movlps || ins == INS_movhpd || ins == INS_movhps || ins == INS_movss ||
-            ins == INS_movlhps || ins == INS_sqrtss || ins == INS_sqrtsd || ins == INS_cvtss2sd || ins == INS_cvtsd2ss);
+    switch (ins)
+    {
+        case INS_cvtsd2ss:
+        case INS_cvtss2sd:
+        case INS_movhpd:
+        case INS_movhps:
+        case INS_movlhps:
```

Comment:
It would be good to move this to `DstDstSrc`. It is incorrectly listed as `DstSrcSrc` and will actually generate an incorrect result on non-AVX machines (see https://github.com/dotnet/coreclr/pull/15538/commits/d413ae72a773b5049ca031ac85741cce18083549#diff-a51165bc1fd2c0fa3987370e2e714356L110).

CC. @fiigii for comment.

Reply:
> It would be good to move this to DstDstSrc. It is incorrectly listed as DstSrcSrc and will actually generate an incorrect result on non-AVX machines

That seems a bug/issue we should fix in the codgen/emitter, rather than change it to `DstDstSrc`.

Reply:
Is it the only one problematic instruction?

Reply:
I believe it is only a bug because it is incorrectly reported as one vs the other.

We switch the operand the is encoded based on `IsDstDstSrc` or `IsDstSrcSrc`:
* https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L9208
* https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L10761

Reply:
That is, the current codegen either duplicates `Dst` or duplicates `Src`.

Duplicating `Src` is incorrect, since the source operands need to be different for `movhlps` (Dst = targetReg, `Src1 = op1Reg`, `Src2 = op1Reg`)

Duplicating `Dst` is okay, since it will end up as `Dst = targetReg = Src1 = op1Reg`, `Src2 = op2Reg`

Reply:
> Is it the only one problematic instruction?

Yes.

All of the other instructions come from APIs which take a single operand, and for which the src should be duplicated.

Reply:
@fiigii @tannergooding My understanding is that consesus reached is to move instruction to DstDstSrc. My opinion based on explanation is the same.

Reply:
Ah, are you trying to implement `Sse.MoveLowToHigh`?
GCC never generates `movlhps` for this intrinsic https://godbolt.org/g/UPBHak

Reply:
> GCC never generates movlhps for this intrinsic https://godbolt.org/g/UPBHak
Clang doesn't, but GCC and MSVC don't appear to.

I am of the opinion that, for any given intrinsic (except for the helper intrinsics) we should be emitting exactly the instruction that maps to the one requested (that is, `MoveLowToHigh` should emit `movlhps`, as is indicated by the Intrinsics Guide, etc: https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&text=movlhps&expand=3608).

An analyzer can suggest a different intrinsic which is better suited or may result in better codegen (ex: Some shuffle instructions with a particular encoding perform the same thing as the unpack instructions, but with a smaller encoding).

Reply:
Ok, I still want to make sure that this change will not impact the current code base. Some `Vector<T>` intrinsics (convert ulong to double) are using `movlhps` with non-AVX environments.

Reply:
I did a quick look as well. It looks like, currently, we are only emitting `movlhps` on non-AVX (as you indicated) and we are using the same reg each time:
`inst_RV_RV(INS_movlhps, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));`, so it wouldn't hit this issue.

Reply:
> so it wouldn't hit this issue.

That would be good.

Reply:
@fiigii @tannergooding 

If we are not hitting this issue now are we committing as is or fix `movlhps` before committing?

Reply:
Because it is a known bug that could impact anyone else attempting use the instruction.

I'm not saying it strictly needs to be part of this PR, just that it would be nice to fix the issue while we are touching this code.

Reply:
> I did a quick look as well. It looks like, currently, we are only emitting movlhps on non-AVX (as you indicated) and we are using the same reg each time:
inst_RV_RV(INS_movlhps, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));

I just recalled that this change may impact AVX/non-AVX2 platforms (e.g., Sandy Bridge).

Please change `movlhps` and run the tests with `set COMPlus_EnableAVX2=0` (debug build CoreCLR).

Reply:
> Please change movlhps and run the tests with set COMPlus_EnableAVX2=0 (debug build CoreCLR).

I believe I had tested this already in https://github.com/dotnet/coreclr/pull/15538.

That being said, I will explicitly run the tests again and collect dumps/disasm

Reply:
> it strictly needs to be part of this PR, just that it would be nice to fix the issue while we are touching this code.

It is my preference as well, let's fix it once we work on this code

Reply:
@fiigii, looks like the exact conditions to get it to repro is a bit complicated.

Basically, you need to be on an AVX/AVX2 machine and call `emitIns_R_R` instead of `emitIns_R_R_R`. This could be triggered by the `UseVEXEncoding() && (reg1 != reg)` check (see https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L4802), which I simplified to just be `UseVEXEncoding()` in my PR.

Once on that path, you will eventually get to `emitOutputRR` which adjusts which register is encoded as the first source based on `IsDstDstSrc` or `IsDstSrcSrc` (see https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L9208). Since `movlhps` is listed as `DstSrcSrc`, it results in `movhlps reg1, reg2, reg2` rather than `movhlps reg1, reg1, reg2` (which is the correct encoding and is emitted for the DstDstSrc case).

It isn't an issue for SIMD today since they just use `emitIns_R_R(tmpReg, tmpReg)` and always get `movhlps tmpReg, tmpReg, tmpReg`.

It isn't an issue for `Sse.MoveHighToLow` since I removed the simplified the check to just always call `emitIns_R_R_R` on VEX machines.

It could potentially be an issue for any future code that just calls `emitIns_R_R` where op1Reg and op2Reg differ.

Reply:
`movlhps` is getting added with the `SSE` intrinsics PR and I have made sure to add it to the correct location.

Since it doesn't exist yet, it doesn't represent any potential issue.